src/services/user.ts

```diff
@@ -0,0 +1,232 @@
+        .where('deactivatedAt', null);
+      return {
+        errCode: -1,
+        data: user,
+      }
+    } catch (err) {
+      console.error(`Error get all users: ${err}`);
+      return userFetchFail();
+    }
+  }
+
+  /**
+   * Delete a user
+   * @param username 
+   * @param soft 
+   */
+  static async deleteByUsername(username: string, soft: boolean) {
+    // Assume a login-ed user tries to delete his/her own acct, no existence check required
+    try {
+      return knex.transaction(async (trx) => {
+        const userQuery = knex<User>(USER);
+        if (soft) {
+          userQuery.where('name', username).update({ deactivatedAt: knex.raw('CURRENT_TIMESTAMP') }).transacting(trx);
+        } else {
+          userQuery.delete().where('name', username).transacting(trx);
+        }
+        const data = await userQuery;
+        console.log(data);
+        // delete follow info
+        const [ { id } ] = await knex<User>(USER).select('id').where('name', username);
+        await knex(USERMAPPING).delete().where('userId', id).orWhere('followerId', id);
+        return {
+          errCode: -1,
+  
+        }
+      });
+
+    } catch (err) {
+      console.error(`Error on deleting a user: ${err}`);
+      return userDeletionFail();
+    }
+  }
+/**
+ * Update user's data by username
+ * @param username 
+ * @param payload 
+ */
+  static async updateByUsername(username: string, payload: Partial<User>) {
+    const userQuery = knex<User>(USER);
+    try {
+      const updateRes = await userQuery.update(payload).where('name', username);
+      if (!!updateRes) {
+        const { data } = await this.findByUsername(username) as SuccessRes;
+        return {
+          errCode: -1,
+          data
+        }
+      } else {
+        throw new Error('');
+      }
+    } catch (err) {
+      console.error(`Error on updating user info: ${err}`);
+      return userUpdateFail();
+    }
+  }
+/**
+ * Update user's password
+ * @param username 
+ * @param oldPassword 
+ * @param newPassword 
+ */
+  static async updatePasswordByUsername(username: string, oldPassword: string, newPassword: string): Promise<SuccessRes | ErrorRes> {
+
+    try {
+      // Toggle when equality check is not implemented in somewhere else
+      // if (oldPassword === newPassword) return pwdEqual
+      const [currPassword] = await knex<User>(USER).select('password').where('name', username).andWhere('deactivatedAt', null);
+      if (!currPassword) return userNotFound();
+      await knex<User>(USER).update({ password: Secure.encrypt(newPassword) }).where('name', username).andWhere('deactivatedAt', null);
+      return {
+        errCode: -1,
+      }
+    } catch (err) {
+      console.error(`Error updating password by username: ${err}`);
+      return userChangePasswordFail();
+    }
+
+  }
+  /**
+   * Get followers of a user
+   * @param username 
+   */
+  static async getFollowerByUsername(username: string): Promise<SuccessRes | ErrorRes> {
```

Comment:
what will happen if one user has many followers (e.g. https://weibo.com/shanghaicity has 9 million followers)? how to solve the problem?

Reply:
did you targeting the 'displaying follower' feature for this scenario? 
if so, the first thought that came across is to expand the request body for additional info like 
{
 where : {},
 sort: {},
 offset: 
 limit: 
} 
to limit the displaying to a certain range.

Reply:
The query may exhaust database or web server's hardware resources without restrictions.
What index should be created if this api supports `where` & `sort` you mentioned above?

Reply:
common 'where- ed' columns depend on the actual user stories.
might need some test ahead for Selectivity value > 0.9 (ideally) as an index. 
if not then combine where-ed columns as a composite index.

db change might happen as well, like split address to street, city, state, dob to y m d, but ideally would be to designed in advance with pre-planned user stories.